modules/dnn/src/darknet/darknet_io.cpp

```diff
@@ -488,9 +532,53 @@ namespace cv {
                 return dst;
             }
 
+            std::vector< std::vector<cv::Mat> > getBlobs(std::istream &ifile, size_t const weights_size,
+                int sizes_weights[], int num, int dim, bool use_batch_normalize)
+            {
+                std::vector< std::vector<cv::Mat> > blobs;
+
+                cv::Mat weightsBlob;
+                weightsBlob.create(dim, sizes_weights, CV_32F);
+                CV_Assert(weightsBlob.isContinuous());
+
+                cv::Mat meanData_mat(1, num, CV_32F);	// mean
+                cv::Mat stdData_mat(1, num, CV_32F);	// variance
+                cv::Mat weightsData_mat(1, num, CV_32F);// scale
+                cv::Mat biasData_mat(1, num, CV_32F);	// bias
+
+                ifile.read(reinterpret_cast<char *>(biasData_mat.ptr<float>()), sizeof(float)*num);
+                if (use_batch_normalize) {
+                    ifile.read(reinterpret_cast<char *>(weightsData_mat.ptr<float>()), sizeof(float)*num);
+                    ifile.read(reinterpret_cast<char *>(meanData_mat.ptr<float>()), sizeof(float)*num);
+                    ifile.read(reinterpret_cast<char *>(stdData_mat.ptr<float>()), sizeof(float)*num);
+                }
+                ifile.read(reinterpret_cast<char *>(weightsBlob.ptr<float>()), sizeof(float)*weights_size);
+
+                // set layer weights
+                std::vector<cv::Mat> layer_blobs;
+                layer_blobs.push_back(weightsBlob);
+                if (!use_batch_normalize) {
+                    // use BIAS in any case
+                    layer_blobs.push_back(biasData_mat);
+                }
+                blobs.push_back(layer_blobs);
+
+                // set batch normalize (mean, variance, scale, bias)
+                if (use_batch_normalize) {
+                    std::vector<cv::Mat> bn_blobs;
+                    bn_blobs.push_back(meanData_mat);
+                    bn_blobs.push_back(stdData_mat);
+                    bn_blobs.push_back(weightsData_mat);
+                    bn_blobs.push_back(biasData_mat);
+                    blobs.push_back(bn_blobs);
+                }
+                return blobs;
+            }
```

Comment:
Looks like it is not so much better. Do you mind to try to unite `convolutional` and `connected` at the previous position?

```cpp
                for (it_type i = net->layers_cfg.begin(); i != net->layers_cfg.end(); ++i) {
                    ++darknet_layers_counter;
                    ++cv_layers_counter;
                    std::map<std::string, std::string> &layer_params = i->second;
                    std::string layer_type = layer_params["type"];

                    if (layer_type == "convolutional" || layer_type == "connected")
                    {
```
than using a different logic for number of weights and their shape